Feature: Lua-first scheduled events and calendar system

## Summary

Add a Lua-first scheduled events / calendar system for Moongate.

The goal is to let shard scripts define recurring or one-shot scheduled events, with runtime execution backed by `TimerWheelService` and a global Lua callback:

- `on_scheduled_event(event)`

## Direction

Lua-first definitions, not JSON-first.

Example shape:

```lua
local scheduled_events = require("common.scheduled_events")

return scheduled_events.event("town_crier_morning", {
    trigger_name = "town_crier_announcement",
    recurrence = "daily",
    time = "09:00",
    time_zone = "Europe/Rome",
    payload = {
        message = "The Britannia bank is now open."
    }
})
```

Then globally:

```lua
function on_scheduled_event(event)
    if event.trigger_name == "town_crier_announcement" then
        log.info("Scheduled event fired: " .. tostring(event.event_id))
    end
end
```

## V1 Scope

- definitions authored in Lua under `moongate_data/scripts/events/**`
- recurrence support:
  - `once`
  - `daily`
  - `weekly`
  - `monthly`
- runtime scheduler above `TimerWheelService`
- publish `ScheduledEventTriggeredEvent` on the game event bus
- global Lua callback `on_scheduled_event(event)`
- small stable payload:
  - `event_id`
  - `trigger_name`
  - `scheduled_at_utc`
  - `fired_at_utc`
  - `recurrence_type`
  - `payload`

## Why

This gives us a clean base for:

- town crier announcements
- holiday events
- world rotations
- scripted scheduled triggers
- future shard automation without hardcoding everything in C#

Implemented the Lua-first scheduled events v1 in .

Delivered:
-  Lua module and  helper
- definition registry + validation for , , , 
- runtime scheduler on top of 
- global 
- explicit Lua callback mapping to 
-  support for recurring events
- docs under 
- tests covering definitions, module helper, scheduler runtime, and script bridge

Verification used:
-   Determining projects to restore...
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Generators/Moongate.Generators.csproj : warning NU1603: Moongate.Generators depends on Microsoft.CodeAnalysis.Analyzers (>= 5.0.0) but Microsoft.CodeAnalysis.Analyzers 5.0.0 was not found. Microsoft.CodeAnalysis.Analyzers 5.3.0 was resolved instead. [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/tests/Moongate.Tests/Moongate.Tests.csproj]
  All projects are up-to-date for restore.
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Generators/Moongate.Generators.csproj : warning NU1603: Moongate.Generators depends on Microsoft.CodeAnalysis.Analyzers (>= 5.0.0) but Microsoft.CodeAnalysis.Analyzers 5.0.0 was not found. Microsoft.CodeAnalysis.Analyzers 5.3.0 was resolved instead.
  Moongate.Generators -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Generators/bin/Debug/netstandard2.0/Moongate.Generators.dll
  Moongate.Core -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Core/bin/Debug/net10.0/Moongate.Core.dll
  Moongate.Server.Metrics -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server.Metrics/bin/Debug/net10.0/Moongate.Server.Metrics.dll
  Moongate.Abstractions -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Abstractions/bin/Debug/net10.0/Moongate.Abstractions.dll
  Moongate.Network -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Network/bin/Debug/net10.0/Moongate.Network.dll
  Moongate.Email -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Email/bin/Debug/net10.0/Moongate.Email.dll
  Moongate.Scripting -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Scripting/bin/Debug/net10.0/Moongate.Scripting.dll
  Moongate.UO.Data -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.UO.Data/bin/Debug/net10.0/Moongate.UO.Data.dll
  Moongate.Network.Packets -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Network.Packets/bin/Debug/net10.0/Moongate.Network.Packets.dll
  Moongate.Stress -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/tools/Moongate.Stress/bin/Debug/net10.0/Moongate.Stress.dll
  Moongate.Persistence -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Persistence/bin/Debug/net10.0/Moongate.Persistence.dll
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(59,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(102,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(110,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'TryEnqueueBookAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(59,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(102,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(110,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'TryEnqueueBookAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(149,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(149,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
  Moongate.Server -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/bin/Debug/net10.0/Moongate.Server.dll
  bin/Debug/net10.0/Moongate.Server: replacing existing signature
  Moongate.Tests -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/tests/Moongate.Tests/bin/Debug/net10.0/Moongate.Tests.dll
Test run for /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/tests/Moongate.Tests/bin/Debug/net10.0/Moongate.Tests.dll (.NETCoreApp,Version=v10.0)
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:  1382, Skipped:     0, Total:  1382, Duration: 7 s - Moongate.Tests.dll (net10.0) -> 1382/1382 passing
-   Determining projects to restore...
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Generators/Moongate.Generators.csproj : warning NU1603: Moongate.Generators depends on Microsoft.CodeAnalysis.Analyzers (>= 5.0.0) but Microsoft.CodeAnalysis.Analyzers 5.0.0 was not found. Microsoft.CodeAnalysis.Analyzers 5.3.0 was resolved instead. [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
  All projects are up-to-date for restore.
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Generators/Moongate.Generators.csproj : warning NU1603: Moongate.Generators depends on Microsoft.CodeAnalysis.Analyzers (>= 5.0.0) but Microsoft.CodeAnalysis.Analyzers 5.0.0 was not found. Microsoft.CodeAnalysis.Analyzers 5.3.0 was resolved instead.
  Moongate.Core -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Core/bin/Release/net10.0/Moongate.Core.dll
  Moongate.Generators -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Generators/bin/Release/netstandard2.0/Moongate.Generators.dll
  Moongate.Abstractions -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Abstractions/bin/Release/net10.0/Moongate.Abstractions.dll
  Moongate.Network -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Network/bin/Release/net10.0/Moongate.Network.dll
  Moongate.Server.Metrics -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server.Metrics/bin/Release/net10.0/Moongate.Server.Metrics.dll
  Moongate.Email -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Email/bin/Release/net10.0/Moongate.Email.dll
  Moongate.Scripting -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Scripting/bin/Release/net10.0/Moongate.Scripting.dll
  Moongate.UO.Data -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.UO.Data/bin/Release/net10.0/Moongate.UO.Data.dll
  Moongate.Network.Packets -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Network.Packets/bin/Release/net10.0/Moongate.Network.Packets.dll
  Moongate.Persistence -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Persistence/bin/Release/net10.0/Moongate.Persistence.dll
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(59,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(102,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(110,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'TryEnqueueBookAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(59,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(102,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(110,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'TryEnqueueBookAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(149,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(149,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
  Moongate.Server -> /Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/bin/Release/net10.0/Moongate.Server.dll

Build succeeded.

/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Generators/Moongate.Generators.csproj : warning NU1603: Moongate.Generators depends on Microsoft.CodeAnalysis.Analyzers (>= 5.0.0) but Microsoft.CodeAnalysis.Analyzers 5.0.0 was not found. Microsoft.CodeAnalysis.Analyzers 5.3.0 was resolved instead. [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Generators/Moongate.Generators.csproj : warning NU1603: Moongate.Generators depends on Microsoft.CodeAnalysis.Analyzers (>= 5.0.0) but Microsoft.CodeAnalysis.Analyzers 5.0.0 was not found. Microsoft.CodeAnalysis.Analyzers 5.3.0 was resolved instead.
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(59,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(102,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(110,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'TryEnqueueBookAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(59,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(102,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(110,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'TryEnqueueBookAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(149,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Services/Items/ItemInteractionService.cs(149,19): warning CA2016: Forward the 'cancellationToken' parameter to the 'PublishAsync' method or pass in 'CancellationToken.None' explicitly to indicate intentionally not propagating the token (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2016) [/Users/squid/projects/personal/moongatev2/.worktrees/scheduled-events/src/Moongate.Server/Moongate.Server.csproj]
    10 Warning(s)
    0 Error(s)

Time Elapsed 00:00:06.70 -> passing

--- a/src/Moongate.Server/Services/Network/NetworkService.cs
+++ b/src/Moongate.Server/Services/Network/NetworkService.cs
@@ -150,7 +150,11 @@ public sealed class NetworkService : INetworkService, ISquidStdService, IAsyncDi
     {
         var session = _sessions.GetOrCreate(e.Client);
         _logger.Information("Client connected: {SessionId}", e.Client.SessionId);
-        _eventBus.Publish(new SessionCreatedEvent(session));
+
+        // Session lifecycle subscribers mutate world state (the spatial index), so their event must be
+        // published on the game loop — like inbound packets (see OnDataReceived) — never on this
+        // transport thread.
+        _mainThreadDispatcher.Post(() => _eventBus.Publish(new SessionCreatedEvent(session)));
     }
 
     private void OnClientDisconnect(object? sender, SquidStdTcpClientEventArgs e)
@@ -158,7 +162,11 @@ public sealed class NetworkService : INetworkService, ISquidStdService, IAsyncDi
         if (_sessions.TryGet(e.Client.SessionId, out var session))
         {
             _sessions.Remove(e.Client.SessionId);
-            _eventBus.Publish(new SessionDestroyedEvent(session));
+
+            // Publish on the game loop, not this transport thread: SpatialSubscriber removes the character
+            // from the spatial index here, and that world mutation must be loop-affine to avoid racing the
+            // loop's own reads/writes (movement re-indexing, range queries).
+            _mainThreadDispatcher.Post(() => _eventBus.Publish(new SessionDestroyedEvent(session)));
         }
 
         _logger.Information("Client disconnected: {SessionId}", e.Client.SessionId);

--- a/tests/Moongate.Tests/Server/LoginFlowIntegrationTests.cs
+++ b/tests/Moongate.Tests/Server/LoginFlowIntegrationTests.cs
@@ -1,4 +1,5 @@
 using System.Buffers.Binary;
+using System.Collections.Concurrent;
 using System.Diagnostics;
 using System.Net;
 using System.Net.Sockets;
@@ -60,6 +61,49 @@ public class LoginFlowIntegrationTests
 
         public void Post(Action action)
             => action();
+    }
+
+    // A stand-in game loop: drains posted work on one dedicated thread and exposes its id, so a test can
+    // prove that work meant for the loop actually ran there and not on the transport thread that posted it.
+    private sealed class LoopThreadDispatcher : IMainThreadDispatcher, IDisposable
+    {
+        private readonly BlockingCollection<Action> _queue = new();
+        private readonly Thread _thread;
+        private readonly ManualResetEventSlim _ready = new();
+
+        public int LoopThreadId { get; private set; }
+
+        public LoopThreadDispatcher()
+        {
+            _thread = new Thread(Run) { IsBackground = true, Name = "test-game-loop" };
+            _thread.Start();
+            _ready.Wait();
+        }
+
+        public int PendingCount => _queue.Count;
+
+        public int DrainPending(double? budgetMs = null)
+            => 0;
+
+        public void Post(Action action)
+            => _queue.Add(action);
+
+        private void Run()
+        {
+            LoopThreadId = Environment.CurrentManagedThreadId;
+            _ready.Set();
+
+            foreach (var action in _queue.GetConsumingEnumerable())
+            {
+                action();
+            }
+        }
+
+        public void Dispose()
+        {
+            _queue.CompleteAdding();
+            _thread.Join(TimeSpan.FromSeconds(2));
+        }
     }
 
     [Fact]
@@ -887,6 +931,46 @@ public class LoginFlowIntegrationTests
         }
     }
 
+    // A client disconnect is raised on the transport thread, but its SessionDestroyedEvent subscribers
+    // mutate world state (SpatialSubscriber removes the character from the spatial index), so they must
+    // run on the game loop — never on the transport thread. This drives a real connect/disconnect through
+    // a dedicated-loop dispatcher and asserts the event surfaces on the loop thread.
+    [Fact]
+    public async Task SessionDestroyed_IsPublishedOnTheGameLoopThread_NotTheTransportThread()
+    {
+        var config = LoopbackConfig();
+        var eventBus = new EventBusService();
+        using var loop = new LoopThreadDispatcher();
+
+        using var destroyed = new ManualResetEventSlim();
+        var subscriberThreadId = 0;
+        eventBus.Subscribe<SessionDestroyedEvent>(
+            (_, _) =>
+            {
+                subscriberThreadId = Environment.CurrentManagedThreadId;
+                destroyed.Set();
+
+                return Task.CompletedTask;
+            }
+        );
+
+        var network = await StartServerAsync(config, eventBus, loop);
+
+        try
+        {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+            socket.Connect(IPAddress.Loopback, network.Port);
+            socket.Close();
+
+            Assert.True(destroyed.Wait(TimeSpan.FromSeconds(2)), "SessionDestroyedEvent was not published.");
+            Assert.Equal(loop.LoopThreadId, subscriberThreadId);
+        }
+        finally
+        {
+            await network.StopAsync(CancellationToken.None);
+        }
+    }
+
     /// <summary>
     /// End-to-end proof that the spatial index and its consumers work together: two real characters log
     /// in and enter the world (same default starting city, so they land on the same tile), the
@@ -1210,6 +1294,22 @@ public class LoginFlowIntegrationTests
 
     private static Task<NetworkService> StartServerAsync(MoongateConfig config)
         => StartServerAsync(config, new EventBusService());
+
+    /// <summary>
+    /// Minimal server (no packet handlers) started with a specific dispatcher — enough to exercise the
+    /// connect/disconnect lifecycle, where the session events are published, without a full login flow.
+    /// </summary>
+    private static async Task<NetworkService> StartServerAsync(
+        MoongateConfig config,
+        IEventBus eventBus,
+        IMainThreadDispatcher dispatcher
+    )
+    {
+        var network = new NetworkService(new SessionManager(), config, [], eventBus, dispatcher);
+        await network.StartAsync(CancellationToken.None);
+
+        return network;
+    }
 
     private static Task<NetworkService> StartServerAsync(MoongateConfig config, IEventBus eventBus)
         => StartServerAsync(config, eventBus, new StubCharacterService(), null);

